### PR TITLE
Support broadcast in native funcol

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119951
* #120070
* #120043
* #120042
* #120025
* #119982
* #119950
* #119881
* __->__ #119229
* #119104

### Summary

@LucasLLC recently implemented `broadcast` in funcol. This is not yet available in the native funcol ops. This PR adds support for broadcast for native funcol.

- Added `_c10d_functional::broadcast` and `_c10d_functional::broadcast_`
- Integrated with python functol broadcast and `AsyncCollectiveTensor`
- Implemented Inductor lowering. Verified correctness and buffer reuse behavior
- Validated dynamo traceability
- Validated AOTInductor compile-ability

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

### CODE REVIEW

File: torch/csrc/distributed/c10d/Functional.cpp [hunk 1]
```diff
@@ -216,6 +215,25 @@ at::Tensor all_to_all_single(
   return output;
 }
 
+at::Tensor& broadcast_(at::Tensor& input, int64_t src, std::string group_name) {
+  c10d::BroadcastOptions opts;
+  opts.rootRank = src;
+  std::vector<at::Tensor> inputs{input};
+
+  auto group = c10d::resolve_process_group(group_name);
+  auto work = group->broadcast(inputs, opts);
```
Comment (wanchaol) [inline torch/csrc/distributed/c10d/Functional.cpp:224]: do we need to check if the `input` is contiguous somewhere for the inplace version?

**Review by wanchaol — APPROVED** [review]
looks great!

**Comment by LucasLLC** [comment]
Cool! Great work @yifuwang !

**Review by LucasLLC — COMMENTED** [review]
Cool! Great work @yifuwang !

File: torch/csrc/distributed/c10d/Functional.cpp (hunk 1 above)
Comment (yifuwang) [reply torch/csrc/distributed/c10d/Functional.cpp:222]: Hey @wanchaol, this is performed in `ProcessGroupNCCL`:

https://github.com/pytorch/pytorch/blob/bc7f3efb0987ed02599a34bd688f04e10a8308a9/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L2975

**Pull request closed without merging**